Jakarta Data DDL Generation enhancements and testing

KyleAure opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Completed following tasks:
- Change DDL file name suffix _repository > _jakartaData
- Change DDL file name to use configDisplayId instead of dataStoreId
- Ensure multiple repos using the same entity and datastore do not cause duplicate statements in ddl files
  - This is already handled by the PersistenceServiceUnit, no changes needed.
- Enable the ability to test the order of SQL statements in ddl files against different databases
- Ensure order of ddl files when multiple repositories use the same datastore

KyleAure committed:
Original commits

DDL file name suffix _repository > _jakartaData

Use configDisplayId instead of dataStoreId

Ensure multiple repos using the same entity and datastore do not cause duplicate statements in ddl files

Enable DB Rotation for testing DDL generation on different databases

Enable the ability to test the order of SQL statements in ddl files against different databases

Ensure order of ddl files when multiple repositories use the same datastore

Allow use of common aliases for database rotation

Revert "Enable DB Rotation for testing DDL generation on different databases"

This reverts commit d5a22815ac2f7e3c751b0755f937979c6abbc92c.

njr-11 commented on dev/io.openliberty.data.internal_fat_datastore/build.gradle:
```diff
@@ -10,5 +10,21 @@
+// If you are preforming database testing use the copyJdbcDrivers
+// task to copy the common JDBC drivers into the
+// publish/shared/resources/jdbc directory
+addRequiredLibraries.dependsOn copyJdbcDrivers
```
The plan that Tracy and I had come up with (which was probably never discussed since transferring the work) was that fat_datastore was supposed to remain Derby only, and fat_ddlgen was supposed to be made to run against the various databases.

tkburroughs replied:
If I recall, the most important difference between the two FATs is that fat_dlgen was supposed to actually execute the DDL to make sure it runs successfully against each database type.  And then the repositories all functioned properly using the generated DDL (and not automatic table creation).  On the other hand fat_datastore basically just checks that the correct table names are in the DDL.  That is because fat_datastore is testing the various DDL file names and combinations of repositories in the same file; whereas fat_ddlgen is focused on the contents of the files, which should include not just create statements, but also updates for constraints.

However, I have no issue with this test also being in the database rotation, so long as we don't loose sight of the need for fat_ddlgen to test more complex tables and actually executing the ddl.

tkburroughs commented on dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java:
```diff
@@ -437,4 +437,11 @@ public String toString() {
+    // Ensures order of DDL generation based on repository interface name
+    @Override
+	@Trivial
+    public int compareTo(FutureEMBuilder o) {
```
Did you consider that each FutureEMBuilder may actually support multiple repositoryInterface types?  This field is set to the first one reported by CDI.  For each repository interface, the code looks for an existing FutureEMBuilder that matches on datastore, app, module, and classloader.  See FutureEMBuilder.equals.

Perhaps FutureEMBuilder should be updated to keep track of all repository interfaces it supports, and then use the one that is alphabetically first?

Or perhaps CDI will always find them in the same order and it doesn't matter?

KyleAure replied:
> Did you consider that each FutureEMBuilder may actually support multiple repositoryInterface types?

I did not realize this, in that case I think the interface field needs to be updated which I have done in my most recent commit and updated the compareTo method to be a bit more pragmatic on sorting.

> Or perhaps CDI will always find them in the same order and it doesn't matter?

I was not able to find any documentation from CDI confirming any sort of order to the `@Observers` annotation.  So to be safe we need to sort here.

tkburroughs reviewed (changes requested):
See my comments about adding to db rotation; I think the most important thing is that fat_ddl gen actually executes the DDL against each database.

My only concern with these changes are my comment about FutureEMBuilders actually representing multiple repository interfaces, but sorting will be based the first one that CDI "discovers".  Is that reliable?

KyleAure committed:
Feedback

njr-11 commented on dev/io.openliberty.data.internal.persistence/resources/io/openliberty/data/internal/persistence/resources/CWWKDMessages.nlsprops:
```diff
@@ -600,7 +600,7 @@ CWWKD1060.name.out.of.scope.useraction=Update the Repository annotation to \
  providing the repository interface.
 
-CWWKD1061.comp.name.in.ejb=CWWKD1061E: The {0} repository interface which is \
+CWWKD1061.comp.name.in.ejb=CWWKD1061E: The {0} repository interfaces which are \
  defined in the {1} enterprise bean module of the {2} application configures \
```
This will also impact several other recently added messages.  Let's not change them in this PR because it isn't clear how best to do so and will further complicate things by requiring an ID review on top of adding merge conflicts.  Note that the change above, if made, would also require switching `configures` to `configure` due to the plural subject.  But for now, you can just skip the message updates by supplying the first interface name as the parameter.

njr-11 commented on dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java:
```diff
@@ -154,13 +156,26 @@ public class FutureEMBuilder extends CompletableFuture<EntityManagerBuilder> imp
+    void addEntity(Class<?> entityClass) {
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-            Tr.debug(this, tc, "add: " + entityClass.getName());
+            Tr.debug(this, tc, "FutureEMBuilder@" + Integer.toHexString(hashCode()) + " addEntity: " + entityClass.getName());
```
Most of this shouldn't have been needed because the `this` parameter ought to have already been printing the hexadecimal hash code of FutureEMBuilder. It should be appearing in the trace if you do the following:
```suggestion
            Tr.debug(this, tc, "addEntity: " + entityClass.getName());
```

njr-11 commented on dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java:
```diff
@@ -154,13 +156,26 @@ public class FutureEMBuilder extends CompletableFuture<EntityManagerBuilder> imp
+    @Trivial
+    void addRepositoryInterface(Class<?> repositoryInterface) {
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+            Tr.debug(this, tc, "FutureEMBuilder@" + Integer.toHexString(hashCode()) + " addRepositoryInterface: " + repositoryInterface.getName());
```
Similar to above, the following should be including the hexadecimal hash code already:
```suggestion
            Tr.debug(this, tc, "addRepositoryInterface: " + repositoryInterface.getName());
```

njr-11 commented on dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java:
```diff
@@ -272,7 +287,7 @@ public EntityManagerBuilder createEMBuilder() {
             throw exc(IllegalArgumentException.class,
                       "CWWKD1061.comp.name.in.ejb",
-                      repositoryInterface.getName(),
+                      repositoryInterfaces,
```
Previously the user would see this:
`The org.example.mycompany.data.repositories.Products repository interface which is defined in the MyEJB enterprise bean module of the MyApp application configures a java:comp value for the dataStore attribute.`

Now they will see:
`The [class org.example.mycompany.data.repositories.Products] repository interfaces which are defined in the MyEJB enterprise bean module of the MyApp application configures a java:comp value for the dataStore attribute.`

At a minimum we should get rid of the extra "class" wording in the message, which would be a trivial change.
After that, there are a couple of directions we can choose from:
1. Make this and the other similar messages plural.
2. Duplicate this and the other similar messages, so that for each there is a single and a plural message to choose from.
3. Keep the singular messages only, and log an error message per repository interface name.
4. Keep the singular messages only, and log one error message for the first repository interface name only.

For this PR, go with 3 or 4 so that it doesn't require an NLS review until after we decide what we really want:
```suggestion
                      repositoryInterfaces.get(0).getName(), // TODO revisit this and other NLS messages
```

KyleAure committed:
Update dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java

Co-authored-by: Nathan Rauh <nathan.rauh@us.ibm.com>

KyleAure committed:
Update dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java

Co-authored-by: Nathan Rauh <nathan.rauh@us.ibm.com>

KyleAure committed:
Update dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java

Co-authored-by: Nathan Rauh <nathan.rauh@us.ibm.com>

KyleAure committed:
Revert to singular message

KyleAure committed:
Resolve merge conflicts

njr-11 commented on dev/com.ibm.ws.persistence.mbean/src/com/ibm/ws/persistence/mbean/internal/DDLGenerationMBeanImpl.java:
```diff
@@ -280,4 +285,37 @@ protected final String getDescription(MBeanOperationInfo opInfo, MBeanParameterI
+                return o1.getClass().getCanonicalName().compareTo(o2.getClass().getCanonicalName());
+            }
+
+            // Now we have to compare two DDLGenerationParticipant implementations of the same type,
+            // but do not provide a comparator. Always put the new participant after the current participant
+            // (o1 > o2) This will maintain insertion order which matches the previous behavior.
+            return 1;
```
It seems like the right solution would be to make DDLGenerationParticipant extend Comparable, forcing all implementations to be comparable.
```suggestion
            // Now we have to compare two DDLGenerationParticipant implementations of the same type,
            // but do not provide a comparator. Always put the new participant after the current participant
            // (o1 > o2) This will maintain insertion order which matches the previous behavior.
            return 1; // TODO avoid this by making DDLGenerationParticipant extend Comparable and update implementations
```

KyleAure replied:
Yes I agree that it would be preferable for all DDLGenerationParticipant to be Comparable.  But I was trying to avoid affecting the current behavior of the other participants.  Adding a TODO here makes sense to consider changing this in the future.

njr-11 commented on dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/service/DBStoreEMBuilder.java:
```diff
@@ -814,13 +825,13 @@ public void generate(Writer out) throws Exception {
     @Override
     public String getDDLFileName() {
-        return databaseStoreId + "_repository";
+        return configDisplayId + "_jakartaData";
```
Would this look better to the user with a capital letter for the spec name?
It would also help make the distinction between a DDL file for a product function vs a technology/specification for the user.
```suggestion
        return configDisplayId + "_JakartaData";
```

KyleAure replied:
Yeah, I think that does look better especially since Jakarta is a proper noun (organization name).

KyleAure committed:
Update dev/com.ibm.ws.persistence.mbean/src/com/ibm/ws/persistence/mbean/internal/DDLGenerationMBeanImpl.java

Co-authored-by: Nathan Rauh <nathan.rauh@us.ibm.com>

KyleAure committed:
Update dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/service/DBStoreEMBuilder.java

Co-authored-by: Nathan Rauh <nathan.rauh@us.ibm.com>

KyleAure committed:
Move and change chase of file names

KyleAure committed:
move and remove temp directory

KyleAure commented:
#libby - check headers
#build - personal build

KyleAure committed:
Fix build error: files are not always sorted on the OS file system

KyleAure commented:
#libby - re-check headers
#build - re-start personal build

KyleAure commented:
#libby - re-check headers
#build - re-build after resolving merge conflicts

KyleAure committed:
fix: merge conflics not previously caught

After resolving merge conflics a local re-run of functional tests failed to
re-compile the source project.  Missed compile time and test time failures.

KyleAure commented:
#libby - re-check headers
#build - re-build after resolving merge conflicts not previously caught 🫨

KyleAure commented:
#libby - re-check headers
#build - re-build after resolving merge conflicts not previously caught 🫨

KyleAure committed:
fix: more merge conflicts

After last round of merge conflicts my local development environment was never fully re-assembled.
This lead to me fixing a test failure which was not necessary.

KyleAure commented:
#libby - re-check headers
#build - personal build

KyleAure commented:
Pipeline: failed with one known defect and one intermittent infrastructure issue
RTC Build: took 10 hours for the bot to respond. 😞
I'll take the passing Pipeline build as enough to merge.
